From: SeongJae Park <sj@kernel.org>
Date: Sun, 22 Dec 2024 12:13:12 -0800
Subject: [PATCH] ==== inclusive damos_filter ====

mm/damon: extend DAMOS filters for inclusion

DAMOS fitlers are exclusive filters.  It only excludes memory of given
criterias from the DAMOS action targets.  This has below limitations.

First, the name is not explicitly explaining the behavior.  This
actually resulted in users' confusions[1].  Secondly, combined uses of
multiple filters provide restriced coverages.  For example, building a
DAMOS scheme that applies the action to memory that belongs to cgroup A
"or" cgroup B is impossible.  A workaround would be using two schemes
that fitlers out memory that not belong to cgroup A and cgroup B,
respectively.  It is cumbersome, and difficult to control quota-like
per-scheme features in orchastration.  Monitoring of filter-passed
memory statistic will also be complicated.

Extend DAMOS filters to support not only exclusion (blocking), but also
inclusion (pass-through) behavior.  For this, add a new damos_filter
struct field called 'pass' for DAMON kernel API users.  The filter works
as an inclusion filter or an exclusion filter when it is set or unset,
respectively.  For DAMON user-space ABI users, add a DAMON sysfs file of
same name under DAMOS filter sysfs directory.  To prevent exposing a
behavioral change to old users, set the blocking as the default
behavior.

Note that pass filters work for only inclusion, not exclusion of memory
that not satisfying the criteria.  And the default behavior of DAMOS for
memory that no filter has involved is that the action can be applied to
those.  This means that installing pass filters without block filters
after them is not useful for some usecases.  Read the design document
update part of this patch series for more details.

[1] https://lore.kernel.org/20240320165619.71478-1-sj@kernel.org

Changes from RFC v2
(https://lore.kernel.org/20241227210819.63776-1-sj@kernel.org)
- Wordsmith messages
- Wordsmith design documentation about monitoring-purpose usage
- Rebase on latest mm-unstable

Changes from RFC v1
(https://lore.kernel.org/20241226221445.78433-1-sj@kernel.org)
- Fix encoding issue on the last patch

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 damon_meta_changes/RRsi2zLl | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 damon_meta_changes/RRsi2zLl

diff --git a/damon_meta_changes/RRsi2zLl b/damon_meta_changes/RRsi2zLl
new file mode 100644
index 000000000000..e69de29bb2d1
-- 
2.39.5

